cpp/src/slice2cpp/Gen.cpp

```diff
@@ -1118,7 +1117,7 @@ Slice::Gen::MetadataVisitor::validate(
         {
             ostringstream ostr;
             ostr << "ignoring invalid metadata '" << *meta << "'";
-            dc->warning(InvalidMetadata, file, line, ostr.str());
+            dc->warning(InvalidMetadata, meta->file(), meta->line(), ostr.str());
```

Comment:
I don't quite follow the `file` vs `meta->file()` in this function. Are they always the same?
Apparently, `file` and `cont` are only used to get `dc` and `dc` is only used to issue warnings.

Reply:
Long story short, it's always safest to get the metadata's file with `meta->file()`.
Most of the time, you're right that it will be the same as `file`, but there's some edge cases where they can differ:
```
// File1
[good-metadata] sequence<int> IntSequence;

// File2
struct S
{
    [bad-metadata] IntSequence is;
}
```
Here, `file` is the file of the actual sequence definition (`File1`), but we want to issue the warning for `File2`.

----

I think we can eliminate the need to get the `dc` with a `file` in the first place, and let `warning` do the lookup using the `file` you pass into it. But obviously, we'd need to move `warning` somewhere else first.
One step at a time though. If you stare at any of our code long enough, you'll find an endless pit of refactoring.